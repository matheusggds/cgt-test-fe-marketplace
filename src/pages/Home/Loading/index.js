import { Grid, Skeleton } from "@mui/material";
import React from "react";

const Loading = () => (
  <Grid container spacing={2}>
    {[0, 0, 0, 0, 0, 0, 0, 0].map((_, idx) => (
      <Grid item md={3} sm={4} xs={6} key={idx}>
        <Skeleton variant="rectangular" height={150} sx={{ mb: 2 }} />
        <Skeleton
          variant="rectangular"
          width={100}
          height={25}
          sx={{ mb: 1 }}
        />
        <Skeleton variant="rectangular" height={20} sx={{ mb: 1 }} />
        <Skeleton variant="rectangular" height={20} sx={{ mb: 1 }} />
        <Skeleton variant="rectangular" height={20} sx={{ mb: 1 }} />
        <Skeleton variant="rectangular" height={20} sx={{ mb: 1 }} />
      </Grid>
    ))}
  </Grid>
);

export default Loading;
