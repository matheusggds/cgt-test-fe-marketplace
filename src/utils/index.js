// convert to dolar string
export const formatToCurrency = (val) =>
  Intl.NumberFormat("en-US", {
    style: "currency",
    currency: "USD",
  }).format(val);
