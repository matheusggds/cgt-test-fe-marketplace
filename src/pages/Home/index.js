import React, { useEffect, useState } from "react";

import axios from "axios";
import { Grid } from "@mui/material";

import ProductCard from "../../components/ProductCard";
import Loading from "./Loading";
import Error from "../../components/Error";

const STATUS = {
  SUCCESS: "success",
  LOADING: "loading",
  ERROR: "error",
};

const Home = () => {
  const [products, setProducts] = useState();
  const [status, setStatus] = useState(STATUS.LOADING);

  useEffect(() => {
    const getProducts = async () => {
      try {
        const { data } = await axios.get("https://dummyjson.com/products");

        console.log("########", data);

        setProducts(data.products);
        setStatus(STATUS.SUCCESS);
      } catch (e) {
        setStatus(STATUS.ERROR);
      }
    };

    getProducts();
  }, []);

  if (status === STATUS.LOADING) {
    return <Loading />;
  }

  if (status === STATUS.ERROR) {
    return <Error />;
  }

  return (
    <Grid container spacing={2}>
      {products.map((el, idx) => (
        <Grid item key={idx} md={3} sm={4} xs={6}>
          <ProductCard data={el} />
        </Grid>
      ))}
    </Grid>
  );
};

export default Home;
