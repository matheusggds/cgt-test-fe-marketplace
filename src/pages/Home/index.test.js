import {
  fireEvent,
  render,
  screen,
  waitFor,
  within,
} from "@testing-library/react";
import { renderWithContext } from "../../utils";

import axios from "axios";

jest.mock("axios");

import Home from "./";

const MOCK = {
  products: [
    {
      id: 1,
      title: "title A",
      description: "description A",
      price: 5,
    },
    {
      id: 2,
      title: "title B",
      description: "description B",
      price: 10,
    },
    {
      id: 3,
      title: "title C",
      description: "description C",
      price: 20,
    },
  ],
};

describe("<Home />", () => {
  describe("View", () => {
    it("should display products correctly", async () => {
      axios.get.mockReturnValue(Promise.resolve({ data: MOCK }));

      const products = renderWithContext(<Home />);

      await screen.findByText("title A");

      expect(products).toMatchSnapshot();
    });

    it("should display loading screen", async () => {
      axios.get.mockReturnValue(Promise.resolve({ data: MOCK }));

      renderWithContext(<Home />);

      await waitFor(() => {
        expect(screen.getByTestId("home-loading")).toBeInTheDocument();
      });
    });

    it("should display error screen", async () => {
      axios.get.mockReturnValue(Promise.reject());

      renderWithContext(<Home />);

      await waitFor(() => {
        expect(
          screen.getByText("Oops! We have a error, try again later")
        ).toBeInTheDocument();
      });
    });
  });

  describe("Requests", () => {
    it("should call API to get products", async () => {
      axios.get.mockReturnValue(Promise.resolve({ data: MOCK }));

      renderWithContext(<Home />);

      waitFor(() => {
        expect(axios.get).toHaveBeenCalled();
        expect(axios.get).toHaveBeenCalledTimes(1);
      });
    });
  });
});
