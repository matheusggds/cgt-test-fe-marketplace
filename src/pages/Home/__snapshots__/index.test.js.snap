// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Home /> View should display products correctly 1`] = `
Object {
  "asFragment": [Function],
  "baseElement": <body>
    <div>
      <div
        class="MuiGrid-root MuiGrid-container MuiGrid-spacing-xs-2 css-mhc70k-MuiGrid-root"
      >
        <div
          class="MuiGrid-root MuiGrid-item MuiGrid-grid-xs-6 MuiGrid-grid-sm-4 MuiGrid-grid-md-3 css-1vm8tgf-MuiGrid-root"
        >
          <div
            class="MuiPaper-root MuiPaper-elevation MuiPaper-rounded MuiPaper-elevation1 MuiCard-root ProductCard css-bhp9pd-MuiPaper-root-MuiCard-root"
          >
            <a
              class="MuiButtonBase-root MuiCardActionArea-root css-1pz8km4-MuiButtonBase-root-MuiCardActionArea-root"
              href="/product/1"
              tabindex="0"
            >
              <img
                alt="title A"
                class="MuiCardMedia-root MuiCardMedia-media MuiCardMedia-img css-o69gx8-MuiCardMedia-root"
                height="140"
              />
              <div
                class="MuiCardContent-root ProductContent css-46bh2p-MuiCardContent-root"
              >
                <div
                  class="MuiTypography-root MuiTypography-h5 MuiTypography-gutterBottom css-h93ljk-MuiTypography-root"
                >
                  title A
                </div>
                <p
                  class="MuiTypography-root MuiTypography-body2 css-r40f8v-MuiTypography-root"
                >
                  description A
                </p>
                <p
                  class="MuiTypography-root MuiTypography-body1 css-1yoa5d8-MuiTypography-root"
                >
                  Price: 
                  <b>
                    $5.00
                  </b>
                </p>
              </div>
              <span
                class="MuiCardActionArea-focusHighlight css-1v2exvi-MuiCardActionArea-focusHighlight"
              />
            </a>
            <div
              class="MuiCardActions-root MuiCardActions-spacing css-1t6e9jv-MuiCardActions-root"
            >
              <button
                class="MuiButtonBase-root MuiButton-root MuiButton-outlined MuiButton-outlinedPrimary MuiButton-sizeSmall MuiButton-outlinedSizeSmall css-1k23hlb-MuiButtonBase-root-MuiButton-root"
                tabindex="0"
                type="button"
              >
                add to cart
              </button>
            </div>
          </div>
        </div>
        <div
          class="MuiGrid-root MuiGrid-item MuiGrid-grid-xs-6 MuiGrid-grid-sm-4 MuiGrid-grid-md-3 css-1vm8tgf-MuiGrid-root"
        >
          <div
            class="MuiPaper-root MuiPaper-elevation MuiPaper-rounded MuiPaper-elevation1 MuiCard-root ProductCard css-bhp9pd-MuiPaper-root-MuiCard-root"
          >
            <a
              class="MuiButtonBase-root MuiCardActionArea-root css-1pz8km4-MuiButtonBase-root-MuiCardActionArea-root"
              href="/product/2"
              tabindex="0"
            >
              <img
                alt="title B"
                class="MuiCardMedia-root MuiCardMedia-media MuiCardMedia-img css-o69gx8-MuiCardMedia-root"
                height="140"
              />
              <div
                class="MuiCardContent-root ProductContent css-46bh2p-MuiCardContent-root"
              >
                <div
                  class="MuiTypography-root MuiTypography-h5 MuiTypography-gutterBottom css-h93ljk-MuiTypography-root"
                >
                  title B
                </div>
                <p
                  class="MuiTypography-root MuiTypography-body2 css-r40f8v-MuiTypography-root"
                >
                  description B
                </p>
                <p
                  class="MuiTypography-root MuiTypography-body1 css-1yoa5d8-MuiTypography-root"
                >
                  Price: 
                  <b>
                    $10.00
                  </b>
                </p>
              </div>
              <span
                class="MuiCardActionArea-focusHighlight css-1v2exvi-MuiCardActionArea-focusHighlight"
              />
            </a>
            <div
              class="MuiCardActions-root MuiCardActions-spacing css-1t6e9jv-MuiCardActions-root"
            >
              <button
                class="MuiButtonBase-root MuiButton-root MuiButton-outlined MuiButton-outlinedPrimary MuiButton-sizeSmall MuiButton-outlinedSizeSmall css-1k23hlb-MuiButtonBase-root-MuiButton-root"
                tabindex="0"
                type="button"
              >
                add to cart
              </button>
            </div>
          </div>
        </div>
        <div
          class="MuiGrid-root MuiGrid-item MuiGrid-grid-xs-6 MuiGrid-grid-sm-4 MuiGrid-grid-md-3 css-1vm8tgf-MuiGrid-root"
        >
          <div
            class="MuiPaper-root MuiPaper-elevation MuiPaper-rounded MuiPaper-elevation1 MuiCard-root ProductCard css-bhp9pd-MuiPaper-root-MuiCard-root"
          >
            <a
              class="MuiButtonBase-root MuiCardActionArea-root css-1pz8km4-MuiButtonBase-root-MuiCardActionArea-root"
              href="/product/3"
              tabindex="0"
            >
              <img
                alt="title C"
                class="MuiCardMedia-root MuiCardMedia-media MuiCardMedia-img css-o69gx8-MuiCardMedia-root"
                height="140"
              />
              <div
                class="MuiCardContent-root ProductContent css-46bh2p-MuiCardContent-root"
              >
                <div
                  class="MuiTypography-root MuiTypography-h5 MuiTypography-gutterBottom css-h93ljk-MuiTypography-root"
                >
                  title C
                </div>
                <p
                  class="MuiTypography-root MuiTypography-body2 css-r40f8v-MuiTypography-root"
                >
                  description C
                </p>
                <p
                  class="MuiTypography-root MuiTypography-body1 css-1yoa5d8-MuiTypography-root"
                >
                  Price: 
                  <b>
                    $20.00
                  </b>
                </p>
              </div>
              <span
                class="MuiCardActionArea-focusHighlight css-1v2exvi-MuiCardActionArea-focusHighlight"
              />
            </a>
            <div
              class="MuiCardActions-root MuiCardActions-spacing css-1t6e9jv-MuiCardActions-root"
            >
              <button
                class="MuiButtonBase-root MuiButton-root MuiButton-outlined MuiButton-outlinedPrimary MuiButton-sizeSmall MuiButton-outlinedSizeSmall css-1k23hlb-MuiButtonBase-root-MuiButton-root"
                tabindex="0"
                type="button"
              >
                add to cart
              </button>
            </div>
          </div>
        </div>
      </div>
    </div>
  </body>,
  "container": <div>
    <div
      class="MuiGrid-root MuiGrid-container MuiGrid-spacing-xs-2 css-mhc70k-MuiGrid-root"
    >
      <div
        class="MuiGrid-root MuiGrid-item MuiGrid-grid-xs-6 MuiGrid-grid-sm-4 MuiGrid-grid-md-3 css-1vm8tgf-MuiGrid-root"
      >
        <div
          class="MuiPaper-root MuiPaper-elevation MuiPaper-rounded MuiPaper-elevation1 MuiCard-root ProductCard css-bhp9pd-MuiPaper-root-MuiCard-root"
        >
          <a
            class="MuiButtonBase-root MuiCardActionArea-root css-1pz8km4-MuiButtonBase-root-MuiCardActionArea-root"
            href="/product/1"
            tabindex="0"
          >
            <img
              alt="title A"
              class="MuiCardMedia-root MuiCardMedia-media MuiCardMedia-img css-o69gx8-MuiCardMedia-root"
              height="140"
            />
            <div
              class="MuiCardContent-root ProductContent css-46bh2p-MuiCardContent-root"
            >
              <div
                class="MuiTypography-root MuiTypography-h5 MuiTypography-gutterBottom css-h93ljk-MuiTypography-root"
              >
                title A
              </div>
              <p
                class="MuiTypography-root MuiTypography-body2 css-r40f8v-MuiTypography-root"
              >
                description A
              </p>
              <p
                class="MuiTypography-root MuiTypography-body1 css-1yoa5d8-MuiTypography-root"
              >
                Price: 
                <b>
                  $5.00
                </b>
              </p>
            </div>
            <span
              class="MuiCardActionArea-focusHighlight css-1v2exvi-MuiCardActionArea-focusHighlight"
            />
          </a>
          <div
            class="MuiCardActions-root MuiCardActions-spacing css-1t6e9jv-MuiCardActions-root"
          >
            <button
              class="MuiButtonBase-root MuiButton-root MuiButton-outlined MuiButton-outlinedPrimary MuiButton-sizeSmall MuiButton-outlinedSizeSmall css-1k23hlb-MuiButtonBase-root-MuiButton-root"
              tabindex="0"
              type="button"
            >
              add to cart
            </button>
          </div>
        </div>
      </div>
      <div
        class="MuiGrid-root MuiGrid-item MuiGrid-grid-xs-6 MuiGrid-grid-sm-4 MuiGrid-grid-md-3 css-1vm8tgf-MuiGrid-root"
      >
        <div
          class="MuiPaper-root MuiPaper-elevation MuiPaper-rounded MuiPaper-elevation1 MuiCard-root ProductCard css-bhp9pd-MuiPaper-root-MuiCard-root"
        >
          <a
            class="MuiButtonBase-root MuiCardActionArea-root css-1pz8km4-MuiButtonBase-root-MuiCardActionArea-root"
            href="/product/2"
            tabindex="0"
          >
            <img
              alt="title B"
              class="MuiCardMedia-root MuiCardMedia-media MuiCardMedia-img css-o69gx8-MuiCardMedia-root"
              height="140"
            />
            <div
              class="MuiCardContent-root ProductContent css-46bh2p-MuiCardContent-root"
            >
              <div
                class="MuiTypography-root MuiTypography-h5 MuiTypography-gutterBottom css-h93ljk-MuiTypography-root"
              >
                title B
              </div>
              <p
                class="MuiTypography-root MuiTypography-body2 css-r40f8v-MuiTypography-root"
              >
                description B
              </p>
              <p
                class="MuiTypography-root MuiTypography-body1 css-1yoa5d8-MuiTypography-root"
              >
                Price: 
                <b>
                  $10.00
                </b>
              </p>
            </div>
            <span
              class="MuiCardActionArea-focusHighlight css-1v2exvi-MuiCardActionArea-focusHighlight"
            />
          </a>
          <div
            class="MuiCardActions-root MuiCardActions-spacing css-1t6e9jv-MuiCardActions-root"
          >
            <button
              class="MuiButtonBase-root MuiButton-root MuiButton-outlined MuiButton-outlinedPrimary MuiButton-sizeSmall MuiButton-outlinedSizeSmall css-1k23hlb-MuiButtonBase-root-MuiButton-root"
              tabindex="0"
              type="button"
            >
              add to cart
            </button>
          </div>
        </div>
      </div>
      <div
        class="MuiGrid-root MuiGrid-item MuiGrid-grid-xs-6 MuiGrid-grid-sm-4 MuiGrid-grid-md-3 css-1vm8tgf-MuiGrid-root"
      >
        <div
          class="MuiPaper-root MuiPaper-elevation MuiPaper-rounded MuiPaper-elevation1 MuiCard-root ProductCard css-bhp9pd-MuiPaper-root-MuiCard-root"
        >
          <a
            class="MuiButtonBase-root MuiCardActionArea-root css-1pz8km4-MuiButtonBase-root-MuiCardActionArea-root"
            href="/product/3"
            tabindex="0"
          >
            <img
              alt="title C"
              class="MuiCardMedia-root MuiCardMedia-media MuiCardMedia-img css-o69gx8-MuiCardMedia-root"
              height="140"
            />
            <div
              class="MuiCardContent-root ProductContent css-46bh2p-MuiCardContent-root"
            >
              <div
                class="MuiTypography-root MuiTypography-h5 MuiTypography-gutterBottom css-h93ljk-MuiTypography-root"
              >
                title C
              </div>
              <p
                class="MuiTypography-root MuiTypography-body2 css-r40f8v-MuiTypography-root"
              >
                description C
              </p>
              <p
                class="MuiTypography-root MuiTypography-body1 css-1yoa5d8-MuiTypography-root"
              >
                Price: 
                <b>
                  $20.00
                </b>
              </p>
            </div>
            <span
              class="MuiCardActionArea-focusHighlight css-1v2exvi-MuiCardActionArea-focusHighlight"
            />
          </a>
          <div
            class="MuiCardActions-root MuiCardActions-spacing css-1t6e9jv-MuiCardActions-root"
          >
            <button
              class="MuiButtonBase-root MuiButton-root MuiButton-outlined MuiButton-outlinedPrimary MuiButton-sizeSmall MuiButton-outlinedSizeSmall css-1k23hlb-MuiButtonBase-root-MuiButton-root"
              tabindex="0"
              type="button"
            >
              add to cart
            </button>
          </div>
        </div>
      </div>
    </div>
  </div>,
  "debug": [Function],
  "findAllByAltText": [Function],
  "findAllByDisplayValue": [Function],
  "findAllByLabelText": [Function],
  "findAllByPlaceholderText": [Function],
  "findAllByRole": [Function],
  "findAllByTestId": [Function],
  "findAllByText": [Function],
  "findAllByTitle": [Function],
  "findByAltText": [Function],
  "findByDisplayValue": [Function],
  "findByLabelText": [Function],
  "findByPlaceholderText": [Function],
  "findByRole": [Function],
  "findByTestId": [Function],
  "findByText": [Function],
  "findByTitle": [Function],
  "getAllByAltText": [Function],
  "getAllByDisplayValue": [Function],
  "getAllByLabelText": [Function],
  "getAllByPlaceholderText": [Function],
  "getAllByRole": [Function],
  "getAllByTestId": [Function],
  "getAllByText": [Function],
  "getAllByTitle": [Function],
  "getByAltText": [Function],
  "getByDisplayValue": [Function],
  "getByLabelText": [Function],
  "getByPlaceholderText": [Function],
  "getByRole": [Function],
  "getByTestId": [Function],
  "getByText": [Function],
  "getByTitle": [Function],
  "queryAllByAltText": [Function],
  "queryAllByDisplayValue": [Function],
  "queryAllByLabelText": [Function],
  "queryAllByPlaceholderText": [Function],
  "queryAllByRole": [Function],
  "queryAllByTestId": [Function],
  "queryAllByText": [Function],
  "queryAllByTitle": [Function],
  "queryByAltText": [Function],
  "queryByDisplayValue": [Function],
  "queryByLabelText": [Function],
  "queryByPlaceholderText": [Function],
  "queryByRole": [Function],
  "queryByTestId": [Function],
  "queryByText": [Function],
  "queryByTitle": [Function],
  "rerender": [Function],
  "unmount": [Function],
}
`;
